src/K4os.Compression.LZ4.Test/PartialDecompressionTests.cs

```diff
@@ -0,0 +1,65 @@
+using System;
+using TestHelpers;
+using Xunit;
+
+namespace K4os.Compression.LZ4.Test;
+
+public class PartialDecompressionTests
+{
+    [Theory]
+    // Full Decompress
+    [InlineData(127, 127)] // prime
+    [InlineData(128, 128)]
+    [InlineData(256, 256)]
+    // Prime Sizes
+    [InlineData(512, 17)] // One prime, one power of two
+    [InlineData(511, 13)] // Both prime
+    [InlineData(511, 31)] // Both prime
+    public void CanDecodePartialCompressedBufferWithSpan(int sourceLength, int numToDecompress)
+    {
+        var targetLength = LZ4Codec.MaximumOutputSize(sourceLength);
+        var source = new byte[sourceLength];
+        var encoded = new byte[targetLength];
+        var decoded = new byte[sourceLength];
+
+        Lorem.Fill(source, 0, source.Length);
+        Fill(encoded, 0xCD);
+        Fill(decoded, 0xCD);
+
+        // Encode a full file.
+        var encodedLength = LZ4Codec.Encode(
+            source.AsSpan(),
+            encoded.AsSpan());
+
+        // Verify data after encodedLength was not overwritten.
+        Check(encoded, encodedLength, encoded.Length - encodedLength, 0xCD);
+
+        // Partially decode the file.
+        var decodedLength = LZ4Codec.DecodePartial(
+            encoded.AsSpan(0, encodedLength),
+            decoded.AsSpan(0, numToDecompress));
+
+        // Verify expected length was decompressed.
+        Assert.Equal(numToDecompress, decodedLength);
+        Check(decoded, 0, decodedLength, source); // Verify correct data was decoded.
+        Check(decoded, decodedLength, sourceLength - decodedLength, 0xCD); // Verify remaining padding is correct.
+    }
+
+    private static void Fill(byte[] buffer, byte value)
+    {
+        for (var i = 0; i < buffer.Length; i++)
+            buffer[i] = value;
+    }
+
+    private void Check(byte[] buffer, int offset, int length, byte value)
+    {
+        for (var i = offset; i < offset + length; i++)
+            Assert.True(buffer[i] == value, $"Value overriden @ {i}");
```

Comment:
Calls to `Assert.True(...)` are slow (for example, because `$"Value overriden @ {i}"` is interpolated anyway creating new unique stream, even if there is no error), and we do this on each byte so at some point it actually cost some CO2.

```
if (buffer[i] != value) Assert.Fail(...)
```

is much faster.

(NOTE: I think XUnit does not actually have `Assert.Fail` and that's why I `throw`)

Reply:
Should be a trivial fix, let me fix this up.

I actually took this routine from another test, hence I kept error the same for consistency.

Reply:
Should be better as of f3e39ee7b2b21702802530744ef5c80e008f8817 .

The .NET BCL doesn't have an implementation of `NotIndexOf` (opposite of `IndexOf`), and I don't know of a library that provides such a routine. This should be trivial to implement, so I might write one for my own [Reloaded.Memory](https://reloaded-project.github.io/Reloaded.Memory/) library sometime.

For now I just resorted to `Assert.Fail`, which does actually exist in xUnit 2, this just throws, so it compiles down to a method call. Technically still not ideal since it's byte by byte and not vectorised, but the inputs aren't that big.